Git/config.sublime-syntax

```diff
@@ -0,0 +1,206 @@
+          pop: true
+        - include: value
+
+  # [section "subsection"]
+  section-header:
+    - match: \[
+      scope: punctuation.definition.section.begin.gitconfig
+      push:
+        - meta_scope: meta.brackets.gitconfig
+        - match: \]
+          scope: punctuation.definition.section.end.gitconfig
+          pop: true
+        - include: illegal-newline  # complain if '[' and ']' on separate lines
+        - include: illegal-slash
+        - include: subsection-name
+        - include: section-name
+    - include: stray-bracket
+
+  # A section is a [section header] and the list of key-value pairs under it.
+  #
+  # [section "subsection"]
+  #   key1 = value1
+  #   key2 = value2
+  section:
+    # [color] subsections are matched separately, because they are the only
+    # sections that take colors for values.
+    - include: section-color  # e.g. [color "branch"], [color "status"]
+    - include: section-other
+
+  # We only want to match color values under these subsections:
+  #   [color "branch"]
+  #   [color "diff"]
+  #   [color "interactive"]
+  #   [color "status"]
+  # Note that this does not match [color] (without the subsection). This is
+  # because [color] doesn't actually take color values.
+  section-color:
+    - match: ^\s*(?=\[\s*color\s+\")
+      push:
+        - match: (?=^\s*\[) # start of new section
+          pop: true
+        - include: section-header
+        - include: key-color-pair
+
+  section-name:
+    - match: '{{section_name}}'
+      scope: entity.name.gitconfig
+
+  # section-other matches all sections except [color "subsection"].
+  section-other:
+    - match: ^\s*(?=\[)
+      push:
+        - match: (?=^\s*\[) # start of new section
+          pop: true
+        - include: section-header
+        - include: key-value-pair
+
+  stray-bracket:
+    - match: \]
+      scope: punctuation.definition.section.end.gitconfig invalid.illegal.stray-bracket.gitconfig
+
+  string:
+    - match: '"'
+      scope: punctuation.definition.string.begin.gitconfig
+      push:
+        - meta_include_prototype: false
+        - meta_scope: string.quoted.double.gitconfig
+        - match: '"'
+          scope: punctuation.definition.string.end.gitconfig
+          pop: true
+        - include: escape
+        - include: illegal-newline  # newlines only permitting after trailing \
+
+  # [section "subsection"]
+  # Subsections are not normal strings. Escapes (e.g. \", \t) and trailing
+  # slashes are not recognized. Newlines are illegal.
+  subsection-name:
+    - match: '"'
+      scope: punctuation.definition.string.begin.gitconfig
+      push:
+        - meta_include_prototype: false
+        - meta_scope: string.quoted.double.gitconfig
+        # '\' is legal in a subsection name, but in front of the trailing '"'
+        # it prevents the '"' from terminating the string. \" is not an escape.
+        # It is still interpreted as two characters.
+        - match: \\"  # no scope needed
```

Comment:
Why is there no constant.character.escape needed here?
```
[x "fo\"o"]
    a = b
```
```
~> git config --get x.fo\"o.a
b
```
Sure, I have to escape it in my shell, but that's a limitation of my particular shell. It'll be passed to git as the (raw) string `x.fo"o.a`. I suggest to add a constant.character.escape scope here.

Reply:
I've been playing around with escapes in subsection names, and discovering some strange behavior. Every character can be "escaped".
```
[x "\f\o\o"]
    a = hello
```
```sh
$ git config --get x.foo.a
hello
```
Traditional escapes like `\n` and `\t` are interpreted as `n` and `t`. I could create a rule to match escapes like this:
```yaml
- match: \\.
  scope: constant.character.escape.gitconfig
```
...but then users will see `\t` matched as an escape and think it's being interpreted as a TAB, instead of `t`. Matching `\t` as an escape would be misleading, but _not_ matching it as an escape makes it look like `\` `t` (two chars). That's also misleading because it's really just `t`.

The only time when `\<char>` is really meaningful is `\"` and `\\`. Every other time, it's equivalent to `<char>`. I can't think of why anyone would put a `\` in a subsection name.

To avoid user confusion, I think we should discourage `\` in subsection names by scoping it as `invalid.illegal`. Here's what I'm doing right now.
```yaml
- match: \\[\\"]
  scope: constant.character.escape.gitconfig
- match: \\
  scope: invalid.unnecessary-escape.gitconfig
```

Reply:
I'm not sure how many color schemes would target `invalid` as a top level scope, as opposed to `invalid.illegal` or `invalid.deprecated`. I think it'd be best to use `invalid.deprecated.unnecessary-escape.gitconfig` here :)